src/model/formulations/GenMorRam2017/startstop.jl

```diff
@@ -0,0 +1,96 @@
+# UnitCommitment.jl: Optimization Package for Security-Constrained Unit Commitment
+# Copyright (C) 2020, UChicago Argonne, LLC. All rights reserved.
+# Released under the modified BSD license. See COPYING.md for more details.
+
+"""
+    _add_startup_shutdown_limit_eqs!(model::JuMP.Model, g::Unit)::Nothing
+
+Startup and shutdown limits from Gentile et al. (2017).
+Eqns. (20), (23a), and (23b) in Kneuven et al. (2020).
+
+Variables
+---
+* :is_on
+* :prod_above
+* :reserve
+* :switch_on
+* :switch_off
+
+Constraints
+---
+* :eq_startstop_limit
+* :eq_startup_limit
+* :eq_shutdown_limit
+"""
+function _add_startup_shutdown_limit_eqs!(model::JuMP.Model, g::Unit)::Nothing
+    # TODO: Move upper case constants to model[:instance]
+    RESERVES_WHEN_START_UP = true
+    RESERVES_WHEN_RAMP_UP = true
+    RESERVES_WHEN_RAMP_DOWN = true
+    RESERVES_WHEN_SHUT_DOWN = true
+
+    eq_startstop_limit = _init(model, :eq_startstop_limit)
+    eq_shutdown_limit = _init(model, :eq_shutdown_limit)
+    eq_startup_limit = _init(model, :eq_startup_limit)
+
+    is_on = model[:is_on]
+    prod_above = model[:prod_above]
+    reserve = model[:reserve]
+    switch_off = model[:switch_off]
+    switch_on = model[:switch_on]
+
+    T = model[:instance].time
+    gi = g.name
+
+    if g.initial_power > g.shutdown_limit
+        #eqs.shutdown_limit[gi, 0] = @constraint(mip, vars.switch_off[gi, 1] <= 0)
+        fix(switch_off[gi, 1], 0.0; force = true)
```

Comment:
I would like to avoid enforcing constraints with `fix` because it's very common to solve the problem with variables fixed, then unfix them. This is used, for example, when computing locational marginal prices (LMP). If we use `fix` as part of the formulation, then the user needs to know which ones to keep, which seems very hard.

Reply:
I think what you are saying is that the user may fix some variables, and then unfix by checking which variables are fixed.  You are worried that the variables that we fix would get changed in the process. This is certainly possible. A solution is for the user to track which variables they have fixed, because they actually have to do that anyway, since the call to `fix` removes bounds (https://jump.dev/JuMP.jl/stable/manual/variables/#Create-a-fixed-variable). For example, in this case, the variable `switch_off` will not be able to be unfixed, unless the user knows the variable's original lower and upper bounds.

I would vote that we have the option to do either a fix or a replace by constant (i.e., keep both in the code). I was using fix because it is much easier to debug a model where the variables are all present, as opposed to trying to figure out how the constant side has been changed. I am adding the "always_create_vars" flag in the `StatusVars` struct, which can be used here too.

Reply:
> I would vote that we have the option to do either a fix or a replace by constant (i.e., keep both in the code).

Sorry that I wasn't clear, but my suggestion was using `@constraint(mip, vars.switch_off[gi, 1] <= 0)` instead of `fix`. I agree that replacing variables by constants makes things more complex.

> A solution is for the user to track which variables they have fixed, because they actually have to do that anyway, since the call to fix removes bounds.

Typically all binary variables (`is_on`, `switch_on`, `switch_off`) are fixed, then unfixed. Currently the user doesn't need to keep track of the original bounds because they are all `[0,1]`. With the changes introduced here, they would have to, otherwise they would get the wrong solution after unfixing. Having a separate constraint avoids this problem more cleanly.

Reply:
I see, that should work. The number of constraints in the model will be artificially much larger, but presolve will handle it. The only issue I see is that it is no longer as easy to directly compare two different formulations on a particular instance based on number of constraints.

Reply:
Do you want me to change the relevant lines where you put a constant to the `@constraint` form? (Probably in a future PR.) If yes, I would change the `always_create_vars` flag to `add_constraint_to_fix_vars` or something like that.

For example:
1. https://github.com/ANL-CEEESA/UnitCommitment.jl/blob/7a1b6f0f55d62e3c3fa0132c3976b489fbfc34b2/src/model/formulations/base/unit.jl#L51
2. https://github.com/ANL-CEEESA/UnitCommitment.jl/blob/dev/src/model/formulations/Gar1962/status.jl#L15-L17

Reply:
Thinking more about this, I think we should always create all decision variables, then fix them using constraints if necessary. This would make the models generated by the package much easier to document, understand and modify. This would also simplify the code. I am fine with entirely removing `always_create_vars` and just making that the default.

If we run into performance issues, we can discuss potential solutions, but I don't think we will, since presolve should handle this, as you mentioned. For statistics, we can either compare the size of the presolved models (probably the best approach in any case), or break down by constraint type.